File: ROID-Forum-Web-Client/Convert Client Code/GMGWebsite/web/Website/Sections/GameSection/ThreadController.dart
library GameThreadController.dart;
import 'ThreadInfo.dart';
import 'GameSection.dart';
import 'dart:html';
import 'CommentInfo.dart';
import 'package:animation/animation.dart';

class ThreadController {
  List<ThreadInfo> threads;
  GameSection sectionController;
  DivElement mainView;
  DivElement headerView;
  DivElement fullView;
  ButtonElement newPostButton, backButton;
  int viewingPosition = 0;
  ThreadInfo viewingThread = null;
  ThreadController(GameSection c) {
    sectionController = c;
    threads = new List();
    mainView = new DivElement();
    headerView = new DivElement();
    SpanElement s = new SpanElement();
    s.text = "The Gamer's Hangout";
    s.style
    ..position = "absolute"
    ..width = "100%"
    ..textAlign = "center"
    ..top = "14px"
    ..fontSize = "25px";
    headerView.append(s);
    fullView = new DivElement();
    makeNewPostButtonComponent();
    makeBackButtonComponent();
    mainView.append(headerView);
  }
  void loggedInEvent() {
      headerView.append(newPostButton);
      for (int i = 0; i < threads.length; i++) {
        ThreadInfo t = threads[i];
        t.fullView.main.append(t.fullView.commentButtonAndBox);
      }
    }
    void loggedOutEvent() {
      newPostButton.remove();
      for (int i = 0; i < threads.length; i++) {
        ThreadInfo t = threads[i];
        t.fullView.commentButtonAndBox.remove();
      }
    }
  void restoreToDefaultState() {
    fullView.remove();
    fullView.children.clear();
    mainView.append(headerView);
    backButton.remove();
    viewingThread = null;
  }
  void threadClicked(ThreadInfo thread) {
    headerView.remove();
    mainView.append(fullView);
    fullView.append(thread.fullView.getDiv());
    mainView.append(backButton);
    viewingThread = thread;
  }
  void backButtonClicked() {
    restoreToDefaultState();
  }
  void updateThreadPositions() {
    for (int i = 0; i < threads.length; i++) {
      ThreadInfo t = threads[i];
      //t.headerView.getDiv().style.top = "${i*85+60}px";
      animate(t.headerView.getDiv(), duration: 250, properties: {'top': "${i*85+60}px"});
    }
  }
  ThreadInfo getThread(int id) {
    for (ThreadInfo t in threads) {
      if (t.getID() == id) return t;
    }
    return null;
  }
  void showThread(int threadID) {
    ThreadInfo thread = getThread(threadID);
    if (thread != null) {
      threadClicked(thread);
    }
  }
  //Actions
  void clearAllThreads() {
    for (int i = 0; i < threads.length; i++) {
      ThreadInfo t = threads[i];
      t.headerView.getDiv().remove();
      t.fullView.getDiv().remove();
    }
    threads.clear();
  }
  void addThread(Map threadMap) {
    int id = threadMap["ID"];
    if (getThread(id) == null) {
    ThreadInfo thread = new ThreadInfo(this);
    thread.setID(id);
    threads.add(thread);

    thread.setOwner(threadMap["Owner"]);
    thread.setTitle(threadMap["Title"]);
    thread.setDescription(threadMap["Description"]);
    if (threadMap["AvatarURL"] != null) {
    thread.setAvatarURL(threadMap["AvatarURL"]);
    }

    List<Map> commentArray = threadMap["Comments"];
    for (int i = 0; i < commentArray.length; i++) {
      Map c = commentArray[i];
      CommentInfo cf = new CommentInfo(thread, this);
      thread.addComment(cf);
      cf.setThreadID(c['ThreadID']);
      cf.setCommentID(c['CommentID']);
      cf.setComment(c['Comment']);
      cf.setOwner(c['Owner']);
      if (c["AvatarURL"] != null) {
        cf.setAvatarURL(c["AvatarURL"]);
      }
    }
    thread.headerView.getDiv().style.opacity = "0.0";
        thread.headerView.getDiv().style.top = "${(threads.length-1)*85+60}px";
        headerView.append(thread.headerView.getDiv());
        animate(thread.headerView.getDiv(), duration: 250, properties: {'opacity': 1.0});
    updateThreadPositions();
    }
  }
  void removeThread(int threadID) {
    ThreadInfo thread = getThread(threadID);
    if (thread != null) {
      threads.remove(thread);
      thread.headerView.getDiv().style.opacity = "1.0";
            animate(thread.headerView.getDiv(), duration: 250, properties: {'opacity': 0.0}).onComplete.listen((_){
              thread.headerView.getDiv().remove();
            });
            thread.fullView.getDiv().style.opacity = "1.0";
                  animate(thread.fullView.getDiv(), duration: 250, properties: {'opacity': 0.0}).onComplete.listen((_){
                    thread.fullView.getDiv().remove();
                  });
      updateThreadPositions();
      if (thread == viewingThread) {
        restoreToDefaultState();
      }
    }
  }
  void updateThread(int threadID, String title, String description) {
    ThreadInfo thread = getThread(threadID);
    if (thread != null) {
      thread.setTitle(title);
      thread.setDescription(description);
    }
  }
  void moveThreadToTop(int threadID) {
    ThreadInfo thread = getThread(threadID);
    if (thread != null) {
      threads.remove(thread);
      threads.insert(0, thread);
      updateThreadPositions();
    }
  }
  void addComment(Map commentMap) {
    int threadID = commentMap["ThreadID"];
    ThreadInfo thread = getThread(threadID);
    if (thread != null) {
      CommentInfo cf = new CommentInfo(thread, this);
      thread.addComment(cf);
      cf.setThreadID(commentMap['ThreadID']);
      cf.setCommentID(commentMap['CommentID']);
      cf.setComment(commentMap['Comment']);
      cf.setOwner(commentMap['Owner']);
      if (commentMap["AvatarURL"] != null) {
        cf.setAvatarURL(commentMap["AvatarURL"]);
      }
    }
  }
  void deleteComment(int threadID, int commentID) {
    print("Handling delete comment");
    ThreadInfo thread = getThread(threadID);
    if (thread != null) {
      thread.removeComment(commentID);
    }
  }
  void updateComment(int threadID, int commentID, String comment) {
    print("Handling update comment");
    ThreadInfo thread = getThread(threadID);
    if (thread != null) {
      CommentInfo ci = thread.getComment(commentID);
      ci.setComment(comment);
    }
  }
  //GUI
  void makeBackButtonComponent() {
    //Add the new post button to it
    backButton = new ButtonElement();
    backButton.style
        ..width = "60px"
        ..height = "30px"
        ..backgroundColor = "white"
        ..borderRadius = "4px"
        ..borderStyle = "solid"
        ..borderWidth = "1px"
        ..borderColor = "#DDDDDD"
        ..fontSize = "15px"
        ..top = "5px"
        ..position = "absolute"
        ..outline = "none"
        ..left = "5px";
    backButton.dataset['active'] = "background: #BBBBBB;";
    backButton.text = 'Back';
    backButton.onClick.listen((var d) {
      backButtonClicked();
    });
  }
  void makeNewPostButtonComponent() {
    //Add the new post button to it
    newPostButton = new ButtonElement();
    newPostButton.style
        ..width = "140px"
        ..height = "40px"
        ..backgroundColor = "white"
        ..borderRadius = "4px"
        ..borderStyle = "solid"
        ..borderWidth = "1px"
        ..borderColor = "#DDDDDD"
        ..fontSize = "15px"
        ..top = "5px"
        ..position = "absolute"
        ..outline = "none"
        ..right = "5px";
    newPostButton.dataset['active'] = "background: #BBBBBB;";
    newPostButton.text = 'Create New Post';
    newPostButton.onClick.listen((var d) {
      sectionController.newPostButtonClicked();
    });
  }
}

File: ROID-Forum-Web-Client/src/Interface/Section/ThreadController.ts
import './ThreadController.scss';

import {ThreadInfo} from "./ThreadInfo";
import {Section} from "./Section";
import {Interface} from "../../Utility/Interface";
import {CommentInfo} from "./CommentInfo";
import {Utility} from "../../Utility/Utility";
import {MessageWriter} from "../../Utility/Message/MessageWriter";
import {SendMessages} from "../../Networking/MessageDefinitions/SendMessages";

type User = { connectionId: string, accountId?: string, displayName?: string };

export class ThreadController {
    threads: ThreadInfo[];
    sectionController: Section;
    mainView: HTMLDivElement;
    headerView: HTMLDivElement;
    fullView: HTMLDivElement;
    newPostButton: HTMLButtonElement;
    backButton: HTMLButtonElement;
    viewingPosition = 0;
    viewingThread: ThreadInfo = null;
    sectionViewingUsers: User[] = [];
    sectionViewersElement: HTMLDivElement;
    threadHeaderTitle: HTMLSpanElement;
    hasDarkTheme: boolean;

    constructor(c: Section, title: string, darkTheme: boolean) {
        this.sectionController = c;
        this.threads = [];
        this.mainView = Interface.Create({
            type: 'div', className: 'ThreadControllerView', elements: [
                this.headerView = Interface.Create({
                    type: 'div', className: 'ThreadHeaderView', elements: [
                        this.threadHeaderTitle = Interface.Create({
                            type: 'span',
                            className: 'ThreadHeaderTitle',
                            text: title
                        }),
                        this.sectionViewersElement = Interface.Create({
                            type: 'span',
                            className: 'SectionViewers',
                            text: ""
                        })
                    ]
                })
            ]
        });
        this.hasDarkTheme = darkTheme;
        if (darkTheme) {
            this.threadHeaderTitle.classList.add('DarkTheme');
            this.sectionViewersElement.classList.add('DarkTheme');
        } else {
            this.threadHeaderTitle.classList.add('LightTheme');
            this.sectionViewersElement.classList.add('LightTheme');
        }

        this.fullView = Interface.Create({type: 'div', className: 'FullView'});
        this.newPostButton = Interface.Create({
            type: 'button',
            className: 'NewPostButton',
            text: 'Create New Post',
            onClick: this.sectionController.newPostButtonClicked
        });
        this.backButton = Interface.Create({
            type: 'button',
            className: 'BackButton',
            onClick: this.backButtonClicked,
            text: 'Back'
        });
    }


    loggedInEvent = () => {
        this.headerView.appendChild(this.newPostButton);
        for (let i = 0; i < this.threads.length; i++) {
            let t: ThreadInfo = this.threads[i];
            t.fullView.main.appendChild(t.fullView.commentButtonAndBox);
        }
    };

    loggedOutEvent = () => {
        this.newPostButton.remove();
        for (let i = 0; i < this.threads.length; i++) {
            let t: ThreadInfo = this.threads[i];
            t.fullView.commentButtonAndBox.remove();
        }
    };

    restoreToDefaultState = () => {
        this.fullView.remove();
        Utility.ClearElements(this.fullView);
        this.mainView.appendChild(this.headerView);
        this.backButton.remove();
        this.setViewingThread(null);
    };

    threadClicked = (thread: ThreadInfo) => {
        this.headerView.remove();
        this.mainView.appendChild(this.fullView);
        this.fullView.appendChild(thread.fullView.getDiv());
        this.mainView.appendChild(this.backButton);
        this.setViewingThread(thread);
    };

    setViewingThread = (thread: ThreadInfo) => {
        if (this.viewingThread != null) {
            let message = new MessageWriter();
            message.addUint8(SendMessages.Controller.Thread);
            message.addString(this.sectionController.sectionId);
            message.addString(this.viewingThread.getThreadId());
            message.addUint8(SendMessages.ThreadMessage.ExitViewingThread);
            this.sectionController.website.networkController.send(message.toBuffer());
        }
        this.viewingThread = thread;
        if (this.viewingThread != null) {
            let message = new MessageWriter();
            message.addUint8(SendMessages.Controller.Thread);
            message.addString(this.sectionController.sectionId);
            message.addString(this.viewingThread.getThreadId());
            message.addUint8(SendMessages.ThreadMessage.BeginViewingThread);
            this.sectionController.website.networkController.send(message.toBuffer());
        }
    };

    backButtonClicked = () => {
        this.restoreToDefaultState();
    };

    updateThreadPositions = () => {
        // Sort the threads by last update time
        this.threads.sort((a, b) => Utility.TimeuuidToMilliseconds(b.getUpdatedTime()) - Utility.TimeuuidToMilliseconds(a.getUpdatedTime()));
        for (let i = 0; i < this.threads.length; i++) {
            let t: ThreadInfo = this.threads[i];
            t.headerView.getDiv().style.top = `${i * 85 + 60}px`;
        }
    };

    getThread(id: string): ThreadInfo {
        for (let t of this.threads) {
            if (t.getThreadId() == id) return t;
        }
        return null;
    }

    showThread = (threadID: string) => {
        let thread: ThreadInfo = this.getThread(threadID);
        if (thread != null) {
            this.threadClicked(thread);
        }
    };

    //Actions
    clearAllThreads = () => {
        for (let i = 0; i < this.threads.length; i++) {
            let t: ThreadInfo = this.threads[i];
            t.headerView.getDiv().remove();
            t.fullView.getDiv().remove();
        }
        this.threads = [];
        this.setViewingThread(null);
    };

    addThread = (threadId: string,
                 creatorAccountId: string,
                 title: string,
                 description: string,
                 createdTime: string,
                 updatedTime: string,
                 commentCount: number,
                 creatorDisplayName: string,
                 creatorAvatarUrl: string) => {
        if (this.getThread(threadId) == null) {
            let thread: ThreadInfo = new ThreadInfo(this, this.hasDarkTheme);
            thread.setThreadId(threadId);
            this.threads.push(thread);

            thread.setCreatorAccountId(creatorAccountId);
            thread.setAvatarURL(creatorAvatarUrl);
            thread.setTitle(title);
            thread.setDescription(description)
            thread.setCreatedTime(createdTime);
            thread.setUpdatedTime(updatedTime);
            thread.setCommentCount(commentCount);
            thread.setCreatorDisplayName(creatorDisplayName);
            thread.headerView.getDiv().style.opacity = "1.0";
            thread.headerView.getDiv().style.top = `${(this.threads.length - 1) * 85 + 60}px`;
            this.headerView.appendChild(thread.headerView.getDiv());

            this.updateThreadPositions();
        }
    };

    removeThread = (threadID: string) => {
        let thread: ThreadInfo = this.getThread(threadID);
        if (thread != null) {
            this.threads.splice(this.threads.indexOf(thread), 1);
            thread.headerView.getDiv().style.opacity = "1.0";
            thread.headerView.getDiv().remove();
            thread.fullView.getDiv().style.opacity = "1.0";
            thread.fullView.getDiv().remove();
            this.updateThreadPositions();
            if (thread == this.viewingThread) {
                this.restoreToDefaultState();
            }
        }
    };

    updateThread = (threadID: string, title: string, description: string) => {
        let thread: ThreadInfo = this.getThread(threadID);
        if (thread != null) {
            thread.setTitle(title);
            thread.setDescription(description);
        }
    };

    updateCommentCount = (threadId: string, commentCount: number) => {
        let thread: ThreadInfo = this.getThread(threadId);
        if (thread != null) {
            thread.setCommentCount(commentCount);
        }
    };

    updateUpdatedTime = (threadId: string, updatedTime: string) => {
        let thread: ThreadInfo = this.getThread(threadId);
        if (thread != null) {
            thread.setUpdatedTime(updatedTime);
            this.updateThreadPositions();
        }
    }

    deleteComment = (threadID: string, commentID: string) => {
        let thread: ThreadInfo = this.getThread(threadID);
        if (thread != null) {
            thread.removeComment(commentID);
        }
    };

    updateComment = (threadID: string, commentID: string, comment: string) => {
        let thread: ThreadInfo = this.getThread(threadID);
        if (thread != null) {
            let ci: CommentInfo = thread.getComment(commentID);
            ci.setComment(comment);
        }
    };

    updateSectionViewersDisplay = () => {
        let currentGuestsOnline = 0;
        let namesOnline = [];
        for (let user of this.sectionViewingUsers) {
            if (user.displayName == undefined) {
                currentGuestsOnline += 1;
            } else {
                namesOnline.push(user.displayName);
            }
        }
        let onlineText = "Currently viewing: ";
        if (namesOnline.length > 0) {
            onlineText += namesOnline.join(", ");
            if (currentGuestsOnline > 0) {
                onlineText += ` and ${currentGuestsOnline} Guest`;
            }
        } else if (currentGuestsOnline > 0) {
            onlineText += `${currentGuestsOnline} Guest`;
        }
        if (currentGuestsOnline > 1) {
            onlineText += "s"
        }
        this.sectionViewersElement.innerText = onlineText;
    }
}